lib/trento/infrastructure/software_updates/suma.ex

```diff
@@ -0,0 +1,188 @@
+defmodule Trento.Infrastructure.SoftwareUpdates.Suma do
+  @moduledoc """
+  SUSE Manager API adapter
+  """
+
+  @behaviour Trento.SoftwareUpdates.Discovery.Gen
+
+  use GenServer, restart: :transient
+
+  alias Trento.Infrastructure.SoftwareUpdates.{Suma.State, SumaApi}
+  alias Trento.SoftwareUpdates
+
+  require Logger
+
+  @default_name "suma"
+
+  @login_retries 5
+
+  def start_link([]), do: start_link(@default_name)
+
+  def start_link(server_name),
+    do: GenServer.start_link(__MODULE__, %State{}, name: process_identifier(server_name))
+
+  @impl true
+  def init(%State{} = state), do: {:ok, state}
+
+  def identify(server_name \\ @default_name),
+    do:
+      server_name
+      |> identificaton_tuple
+      |> :global.whereis_name()
+
+  def setup(server_name \\ @default_name),
+    do:
+      server_name
+      |> process_identifier
+      |> GenServer.call(:setup)
+
+  @impl true
+  def get_system_id(fully_qualified_domain_name, server_name \\ @default_name),
+    do:
+      server_name
+      |> process_identifier
+      |> GenServer.call({:get_system_id, fully_qualified_domain_name})
+
+  @impl true
+  def handle_call(:setup, _from, %State{} = state) do
+    case setup_auth(state) do
+      {:ok, new_state} ->
+        {:reply, :ok, new_state}
+
+      {:error, _} = error ->
+        {:reply, error, state}
+    end
+  end
+
+  @impl true
+  def handle_call(request, from, %State{auth: nil} = state),
+    do: {:noreply, state, {:continue, {:setup, from, request}}}
+
+  @impl true
+  def handle_call(
+        {:get_system_id, fully_qualified_domain_name},
+        _from,
+        %{
+          url: url,
+          auth: auth_cookie
+        } = state
+      ) do
+    {:reply, get_system_id(url, auth_cookie, fully_qualified_domain_name), state}
+  end
+
+  @impl true
+  def handle_continue({:setup, from, previous_message}, %State{} = state) do
+    case setup_auth(state) do
+      {:ok, new_state} ->
+        Process.send(self(), {previous_message, reply_to: from}, [:nosuspend, :noconnect])
+        {:noreply, new_state}
+
+      {:error, reason} ->
+        {:stop, reason}
+    end
+  end
+
+  @impl true
+  def handle_info(
+        {{:get_system_id, fully_qualified_domain_name}, reply_to: from},
+        %State{
+          url: url,
+          auth: auth_cookie
+        } = state
+      ) do
+    GenServer.reply(from, get_system_id(url, auth_cookie, fully_qualified_domain_name))
+
+    {:noreply, state}
+  end
```

Comment:
It's me for sure but I'm not fully understanding why do we need this `handle_info` clause if we have the `handle_call` that reacts to the same message

Reply:
because of the login loop. 
If we call `Suma.get_system_id("foobar")` before any setup was done and no auth is in the process state we need to login first.

As a first step we fall [in this case](https://github.com/trento-project/web/pull/2363/files#diff-9e97d75bb08b31078aef739786dbfd385efcf710824c9a21cddb515a58630354R58)

```elixir
def handle_call(request, from, %State{auth: nil} = state),
    do: {:noreply, state, {:continue, {:setup, from, request}}}
```
continuing [here](https://github.com/trento-project/web/pull/2363/files#diff-9e97d75bb08b31078aef739786dbfd385efcf710824c9a21cddb515a58630354R74) and logging in
```elixir
def handle_continue({:setup, from, previous_message}, %State{} = state) do
  case setup_auth(state) do
    {:ok, new_state} ->
      Process.send(self(), {previous_message, reply_to: from}, [:nosuspend, :noconnect])
      {:noreply, new_state}

    {:error, reason} ->
      {:stop, reason}
  end
end
```
and finally by doing `Process.send(self(), {previous_message, reply_to: from}, [:nosuspend, :noconnect])` we make sure the handle info for get_system_id is executed where we actually reply with `GenServer.reply(...)`

I am not a fan of that either and very keen to refactor. 
I believe we will come back to this service again and if you're ok we can let this sink and refactor as we go. Would that be ok?

Reply:
Thank you so much for the explanation! So far I have no clue about how we could refactor this, so okay to keep it. I was just curious about the flow :+1: